dimod/generators/mimo.py

```diff
@@ -0,0 +1,1058 @@
+        ...     SNRb=SNRb, 
+        ...     F_distribution = ('binary', 'real'))
+
+    .. [#T02] T. Tanaka IEEE TRANSACTIONS ON INFORMATION THEORY, VOL. 48, NO. 11, NOVEMBER 2002
+
+    .. [#R20] J. Raymond, N. Ndiaye, G. Rayaprolu and A. D. King, 
+        "Improving performance of logical qubits by parameter tuning and topology compensation," 
+        2020 IEEE International Conference on Quantum Computing and Engineering (QCE), 
+        Denver, CO, USA, 2020, pp. 295-305, doi: 10.1109/QCE49297.2020.00044.
+
+    .. [#Prince] Various (https://paws.princeton.edu/)
+    """
+
+    random_state = _make_random_state(seed)
+
+    if y is not None:
+        if len(y.shape) == 1:
+            y = y.reshape(y.shape[0], 1)
+        elif len(y.shape) != 2 or y.shape[1] != 1:
+            raise ValueError(f"y should have shape (n, 1) or (n,) for some n; given: {y.shape}")
+
+    if F is None:
+
+        if num_transmitters:
+            if num_transmitters <= 0:
+                raise ValueError('Configured number of transmitters must be positive')
+        elif transmitted_symbols:
+            num_transmitters = len(transmitted_symbols)
+        else:
+           ValueError('`num_transmitters` is not specified and cannot'
+                'be inferred from `F` or `transmitted_symbols` (both None)') 
+           
+        if num_receivers:
+            if num_receivers <= 0:
+                raise ValueError('Configured number of receivers must be positive')
+        elif y:
+            num_receivers = y.shape[0]
+        elif channel_noise:
+            num_receivers = channel_noise.shape[0]  
+        else:
+            raise ValueError('`num_receivers` is not specified and cannot'
+                'be inferred from `F`, `y` or `channel_noise` (all None)')  
+
+        F, channel_power = create_channel(num_receivers=num_receivers,
+                                                num_transmitters=num_transmitters,
+                                                F_distribution=F_distribution,
+                                                random_state=random_state,
+                                                attenuation_matrix=attenuation_matrix)
+    else:
+        channel_power = num_transmitters
+
+    if y is None:
+        y, _, _, _ = _create_signal(F,
+                                    transmitted_symbols=transmitted_symbols,
+                                    channel_noise=channel_noise,
+                                    SNRb=SNRb, modulation=modulation,
+                                    channel_power=channel_power,
+                                    random_state=random_state)
+
+    h, J, offset = _yF_to_hJ(y, F, modulation)
+
+    if use_offset:
+        #NB - in this form, offset arises from
+        return dimod.BQM(h[:, 0], J, 'SPIN', offset=offset)
+    else:
+        np.fill_diagonal(J, 0)
+        return dimod.BQM(h[:, 0], J, 'SPIN')
+
+def spin_encoded_comp(lattice: GraphLike,
```

Comment:
Now I remember why I used ``GraphLike`` rather than ``networkx.Graph``: we don't want to make Networkx a dependency of dimod:

```
 def spin_encoded_comp(lattice: networkx.Graph,
NameError: name 'networkx' is not defined
```

It's a bit more expansive but I think it's the closest we can do here

Reply:
You can also just say:
```
def spin_encoded_comp(lattice: 'networkx.Graph',
```

But the problem with using `GraphLike` here is that we don't actually support it -- i.e. you can specify a lattice by providing an e.g. 5 -- which would mean `K5` as `GraphLike`.

Reply:
Why not https://github.com/dwavesystems/dimod/blob/8433f221a1e79101e1db0d80968ab5a2f59b865d/dimod/typing.py#L76 ?

Reply:
@arcondello, because that function doesn't actually accept `GraphLike`, [but only `nx.Graph`](https://github.com/dwavesystems/dimod/blob/40ae9c2406319133e09007c61eb05167354550aa/dimod/generators/mimo.py#L933).